mrjob/dataproc.py

```diff
@@ -0,0 +1,873 @@
+        """
+        if not self._opts['check_input_paths']:
+            return
+
+        for path in self._input_paths:
+            if path == '-':
+                continue  # STDIN always exists
+
+            if is_uri(path) and not is_gcs_uri(path):
+                continue  # can't check non-GCS URIs, hope for the best
+
+            if not self.fs.exists(path):
+                raise AssertionError(
+                    'Input path %s does not exist!' % (path,))
+
+    def _check_output_not_exists(self):
+        """Verify the output path does not already exist. This avoids
+        provisioning a cluster only to have Hadoop refuse to launch.
+        """
+        if self.fs.exists(self._output_dir):
+            raise IOError(
+                'Output path %s already exists!' % (self._output_dir,))
+
+    def _add_bootstrap_files_for_upload(self):
+        """Add files needed by the bootstrap script to self._upload_mgr.
+
+        Tar up mrjob if bootstrap_mrjob is True.
+
+        Create the master bootstrap script if necessary.
+
+        """
+        # lazily create mrjob.tar.gz
+        if self._bootstrap_mrjob():
+            self._create_mrjob_tar_gz()
+            self._bootstrap_dir_mgr.add('file', self._mrjob_tar_gz_path)
+
+        # all other files needed by the script are already in
+        # _bootstrap_dir_mgr
+        for path in self._bootstrap_dir_mgr.paths():
+            self._upload_mgr.add(path)
+
+        # now that we know where the above files live, we can create
+        # the master bootstrap script
+        self._create_master_bootstrap_script_if_needed()
+        if self._master_bootstrap_script_path:
+            self._upload_mgr.add(self._master_bootstrap_script_path)
+
+    def _add_job_files_for_upload(self):
+        """Add files needed for running the job (setup and input)
+        to self._upload_mgr."""
+        for path in self._get_input_paths():
+            self._upload_mgr.add(path)
+
+        for path in self._working_dir_mgr.paths():
+            self._upload_mgr.add(path)
+
+        # TODO - mtai @ davidmarin - hadoop_streaming_jar is currently ignored, see _HADOOP_STREAMING_JAR_URI
+        # if self._opts['hadoop_streaming_jar']:
+        #     self._upload_mgr.add(self._opts['hadoop_streaming_jar'])
+
+        for step in self._get_steps():
+            if step.get('jar'):
+                self._upload_mgr.add(step['jar'])
+
+    def _upload_local_files_to_gcs(self):
+        """Copy local files tracked by self._upload_mgr to GCS."""
+        log.info('Copying non-input files into %s' % self._upload_mgr.prefix)
+
+        for path, gcs_uri in self._upload_mgr.path_to_uri().items():
+            log.debug('uploading %s -> %s' % (path, gcs_uri))
+
+            # TODO - mtai @ davidmarin - Implement upload function for other FSs
+            self.fs.upload(path, gcs_uri)
+
+    ### Running the job ###
+
+    def cleanup(self, mode=None):
```

Comment:
Lots of comments about cleanup below; bottom line is that writing code to clean up stuff on cloud storage is a very low priority (you *do* want to make sure to shut down clusters).